### PR TITLE
Gelato: ForwardRequest and TaskStatus calls, and EIP-712 signing for ForwardRequests

### PR DESCRIPTION
**Pull request by webbhorn**
[ Posting this for early directional feedback / tips, and acknowledging the TODOs that ought to be resolved before we merge with main. ]

Goal of this PR is to add support for posting ForwardRequest messages to Gelato and obtaining the status of those requests post-submission-to-Gelato.

Unfortunately this is a large PR. If I could suggest a couple of entry points for review, I'd recommend starting with `rust/gelato/src/fwd_req_op.rs`, and in particular the `ForwardRequestOp::run()` method. This is intended to be the entrypoint for usage of this API. An example of its expected usage is in `rust/gelato/src/bin/gelato-cli.rs`, which is just a little demo tool which may or may not even be worth checking in.

The key abstractions introduced in this PR are:

-  `ForwardRequestOp` (in `fwd_req_op.rs`): Encapsulates the overall logic of submitting a message to Gelato's Relay service and waiting for the result. Usage is to create an op, and then invoke `op.run().await()`. The encapsulated logic spans an arbitrary number of RPCs to different Gelato endpoints, with Gelato-specific logic for retrying and error handling.

-  `ForwardRequestCall` (in fwd_req_call.rs): Logic for making a single, individual RPC to the Gelato endpoint responsible for accepting the initial forward request submission, which it then enqueues and handles asynchronously.

-  `fwd_req_sig.rs`, which contains an impl of the  ethers::types::transaction::eip712::Eip712 trait for `ForwardRequestOpArgs`. This allows an ethers::signers::Signer to sign our message before we submit it to the relevant Gelato endpoint.

-  `TaskStatusCall` in `task_status_call.rs`: Analogous to `ForwardRequestCall` in that it encapsulates the logic and state required to make an individual call to a Gelato endpoint to get the status of a previously-submitted request.

### CODE REVIEW

File: rust/gelato/src/fwd_req_op.rs
Comment (webbhorn) [inline rust/gelato/src/fwd_req_op.rs:82]: intended to come back and plumb these all by ref rather than cloning everything, was just done for expediency at first

File: rust/gelato/src/fwd_req_op.rs
Comment (mattiecnvr) [inline rust/gelato/src/fwd_req_op.rs:103]: Might I suggest `for attempt in 0..`

File: rust/gelato/src/fwd_req_op.rs
Comment (mattiecnvr) [inline rust/gelato/src/fwd_req_op.rs:114]: `self.http.clone()` should also work

File: rust/gelato/src/task_status_call.rs
Comment (mattiecnvr) [inline rust/gelato/src/task_status_call.rs:120]: Uh, does this work as expected with the `camelCase` setting? I could see it stripping the `_`

File: rust/gelato/src/chains.rs
Comment (webbhorn) [reply rust/gelato/src/chains.rs:104]: Is our bar for 'ok to just `unwrap`' just 'code mistake'? I get what you're saying, but I wonder if we ought to consider a tighter standard in certain contexts where panics are costly and deterministic and would clearly affect liveness of the network. Here is an example of the category of scenario I'd worry about: an agent is getting lots of traffic to forward messages from chain A to chain B. It has been running for a long time without issue. Suddenly somebody on chain A posts a message to be forwarded to chain C. For the first time, the agent consults rust/gelato/src/chains.rs Chain::from_str() which consults the hardcoded string address. For whatever reason it is (let's say) an odd number of characters due to a copy/paste error, and we didn't/don't have universal line coverage for unit tests, and so it won't parse and we panic. When the agent comes back up, it will find the same work to deliver the message from A to C and panic again, and will do so idnefinitely until we are alerted. That is not ideal in its own right, but more sadly, it would halt transmission of messages from A->B which had been heretofore working smoothly. Wouldn't we prefer to return an error for which we had monitoring that would alert us / the right person as to the condition and have the agent continue processing what it can meanwhile?

File: rust/gelato/src/fwd_req_call.rs
Comment (webbhorn) [reply rust/gelato/src/fwd_req_call.rs:27]: I guess your motivation for that is something like: specifically we want to invalidate the call object after invocation -- if somebody wants to retry the call, they need to make a new one and then call run() on the new one?

I see the following guidance in chapter 5.3 of the rust book which makes some sense to me, just want to make sure I understand the recommenation:

""""Having a method that takes ownership of the instance by using just self as the first parameter is rare; this technique is usually used when the method transforms self into something else and you want to prevent the caller from using the original instance after the transformation."""

File: rust/gelato/src/fwd_req_call.rs
Comment (webbhorn) [reply rust/gelato/src/fwd_req_call.rs:57]: There were enough modifications I had to make to the serde attribute invocation that it seemed more readable to me to write it out explicitly:

-  hoist the two subtypes up to the top-level dict (`#[serde(flatten)]`)
- make sure the integers for gas and maxfee were enclosed within quotes, since they seem to be javascript bignums (or something)
- add a new top-level dict field 'typeId' with const literal value 'ForwardRequest'
- make sure that the sponsor signature field is prefixed with '0x' since it otherwise just looks like a string

If you still think this would be better and more readable even with all of these customizations to the serde() attribute, I'm open to making the change.

File: rust/gelato/src/task_status_call.rs
Comment (webbhorn) [reply rust/gelato/src/task_status_call.rs:120]: It does work. camelCase indeed strips the '_' and that is the intended effect. I can't name the field 'type' because that is a Rust keyword, but the json handed to us in the http response names the field 'type'. Maybe it'd be more idiomatic if I named it something like 'type_field' and did a #[serde_rename] ?

**Comment by webbhorn** [comment]
Thanks for the comments, replied to a few and will make some code modifications before replying to the others.

File: rust/gelato/src/task_status_call.rs
Comment (webbhorn) [reply rust/gelato/src/task_status_call.rs:120]: I've also now discovered rust's escaping capability wherein I could maybe do something like:

...
pub struct BigNumType {
    pub hex: String,
    pub r#type: String,
}

File: rust/gelato/Cargo.toml
Comment (webbhorn) [reply rust/gelato/Cargo.toml:13]: Nope, I think all I need is ["macros"].

File: rust/gelato/src/fwd_req_op.rs
Comment (webbhorn) [reply rust/gelato/src/fwd_req_op.rs:103]: Done.

File: rust/gelato/src/chains.rs
Comment (webbhorn) [reply rust/gelato/src/chains.rs:15]: Thanks. I've made the discriminators match.

File: rust/gelato/src/task_status_call.rs
Comment (mattiecnvr) [reply rust/gelato/src/task_status_call.rs:117]: Yeah, this is one of the only times where I actually can see the ident escaping making sense. So I guess I leave this up to you which option you want to use.

File: rust/gelato/src/fwd_req_call.rs
Comment (mattiecnvr) [reply rust/gelato/src/fwd_req_call.rs:55]: Nope, but please more or less copy this to a doc in the code.

File: rust/gelato/src/chains.rs
Comment (mattiecnvr) [reply rust/gelato/src/chains.rs:104]: Ideally we are testing our code and will run into them there. I see what you are saying but my personal heuristic for panics is:

1) Anything we know for a fact cannot happen (usually best to document why it is impossible if it is not overly obvious to a casual observer)
2) code error that really should just be fixed if it ever occurs for any reason
3) anything that we don't think will actually happen **and** we are in a hurry (this should be an `except` not an `unwrap` so we can pin it down and fix if it turns out to actually happen)

Comment (webbhorn) [reply rust/gelato/src/chains.rs:104]: Those are fine heuristics, but do you have any thoughts about the liveness concern I raised?

Comment (mattiecnvr) [reply rust/gelato/src/chains.rs:104]: The agents crash and restart as it stands, it's not great but that is why I suggest focusing on code errors because then it's like an instant bug report. In this exact case if you think keeping it as you have it is best, I am not giving a hard and fast "make this an unwrap".

File: rust/gelato/src/task_status_call.rs
Comment (tkporter) [inline rust/gelato/src/task_status_call.rs:6]: is this the same for mainnet and testnets?

File: rust/gelato/src/fwd_req_call.rs
Comment (tkporter) [inline rust/gelato/src/fwd_req_call.rs:71]: not sure I follow?

File: rust/gelato/src/fwd_req_op.rs
Comment (tkporter) [inline rust/gelato/src/fwd_req_op.rs:60]: I think this is probably best for a future PR?

**Review by tkporter — COMMENTED** [review]
Looks good! I see there are a lot of TODOs so I haven't done an extremely deep pass around those spots. Would you prefer to merge early and make incremental changes to fix those TODOs or to address some more of them before merging this?

File: rust/gelato/src/fwd_req_op.rs
Comment (webbhorn) [reply rust/gelato/src/fwd_req_op.rs:60]: Do you mean the whole 'op', or just the inclusion of things like DB interface / prometheus integration? I'm trying to cut scope today so that I can go ahead and merge *something*, open to either

File: rust/gelato/src/fwd_req_op.rs
Comment (webbhorn) [reply rust/gelato/src/fwd_req_op.rs]: Resolving this since I've punted fwd_req_op.rs until a future PR.

File: rust/gelato/Cargo.toml
Comment (webbhorn) [reply rust/gelato/Cargo.toml]: Cargo.toml is signicantly trimmed down now. As we've discussed in person, I've made versions match with the other abacus rust Cargo version requirements, or for newly-introduced dependencies, only depend on a major version, unless a package's docs recommend something specific. Gonna resolve this thread, but please feel free to raise again if you want more/other changes in rust/gelato/Cargo.toml.

File: rust/gelato/src/bin/gelato_cli.rs
Comment (webbhorn) [reply rust/gelato/src/bin/gelato_cli.rs]: Thanks. I agree that'd probably be cleaner. In any case I've punted on the CLI demo to a future PR so gonna resolve this.

File: rust/gelato/src/fwd_req_call.rs
Comment (webbhorn) [reply rust/gelato/src/fwd_req_call.rs:117]: It was intentional here, but I've noted the suggestion and am updating elsewhere where it seems like doc-level comments would be useful.

File: rust/gelato/src/fwd_req_call.rs
Comment (webbhorn) [reply rust/gelato/src/fwd_req_call.rs:71]: Done, thanks!

File: rust/gelato/src/task_status_call.rs
Comment (webbhorn) [reply rust/gelato/src/task_status_call.rs:6]: Seemingly, yes. We could ask Ed, but it has worked with all testnets I've tried, and it seems to transparently be the case when reading the relay sdk typescript implementation

File: rust/gelato/src/fwd_req_call.rs
Comment (webbhorn) [reply rust/gelato/src/fwd_req_call.rs:87]: I think this was a stale TODO when I had a more convoluted type conversion originally.

File: rust/gelato/src/fwd_req_op.rs
Comment (tkporter) [reply rust/gelato/src/fwd_req_op.rs]: I meant the inclusion of things like DB / prometheus. Happy to defer to you on including op in this PR or not

File: rust/gelato/src/fwd_req_call.rs
Comment (webbhorn) [reply rust/gelato/src/fwd_req_call.rs:157]: Thanks. Yeah, sorry about the ~/.rustfmt.toml override, didn't intend to send for review with that. Hopefully the rest of these tests look more reasonable now too.

File: rust/gelato/src/chains.rs
Comment (webbhorn) [reply rust/gelato/src/chains.rs:104]: Alright, I think I'm going to leave as-is for now if that's cool (taking your 'not giving a hard and fast' as a signal that it is). One additional reason I'm ok with returning an error (aside from liveness concerns) is that since we handle the default case for an UnknownRelayForwardAddress, any caller has to handle the possibility of an error anyway, so I'm not sure we'd be getting any huge simplicity gain.

File: rust/gelato/src/fwd_req_op.rs
Comment (webbhorn) [reply rust/gelato/src/fwd_req_op.rs]: Thanks, it looks really useful, but punting on this entire module now, so will address in followup PR.

File: rust/gelato/src/fwd_req_call.rs
Comment (webbhorn) [reply rust/gelato/src/fwd_req_call.rs]: Ok I've changed to self, but still wonder about the question I posed a couple weeks ago, so will leave this thread open

File: rust/gelato/src/fwd_req_op.rs
Comment (webbhorn) [reply rust/gelato/src/fwd_req_op.rs]: the Op stuff is in a future PR now

File: rust/gelato/src/fwd_req_op.rs
Comment (webbhorn) [reply rust/gelato/src/fwd_req_op.rs]: Yup whole op is in future PR now so resolving this. Agree DB/prometheus functionality should be introduced separately.

File: rust/gelato/src/chains.rs
Comment (mattiecnvr) [inline rust/gelato/src/chains.rs:1]: Thanks for comment

File: rust/gelato/src/chains.rs
Comment (mattiecnvr) [inline rust/gelato/src/chains.rs:41]: You might like to know that you can `use Chain::*` within the function and avoid having to type as much. I don't always do this, but sometimes it is nice.

File: rust/gelato/src/chains.rs
Comment (mattiecnvr) [inline rust/gelato/src/chains.rs:15]: If I were you I would consider using a one-off (private) macro to prevent needing to list the same values more than once. I can help if you are interested.

File: rust/gelato/src/chains.rs
Comment (mattiecnvr) [inline rust/gelato/src/chains.rs:61]: Would it make sense to also have `TryFrom<u32> for Chain`?

**Review by mattiecnvr — APPROVED** [review]

File: rust/gelato/src/chains.rs
Comment (webbhorn) [reply rust/gelato/src/chains.rs:15]: I have mostly modeled this after the ethers::types::Chain implementation here: https://github.com/gakonst/ethers-rs/blob/master/ethers-core/src/types/chain.rs, which does not use a macro. I don't personally mind listing things twice, if you're ok with it, could we skip the macro for now?

File: rust/gelato/src/chains.rs
Comment (webbhorn) [reply rust/gelato/src/chains.rs:41]: I didn't know, thanks, that's handy. As above, by default I'd just use the same convention as the ethers chains impl in https://github.com/gakonst/ethers-rs/blob/master/ethers-core/src/types/chain.rs if you don't have a strenuous objection

File: rust/gelato/src/chains.rs
Comment (webbhorn) [reply rust/gelato/src/chains.rs:61]: Do you say that just for symmetry purposes? I don't have an active need for it (whereas the other type conversions I do, I just ended up punting the CLI and Op code to a future PR). Even if it's symmetrical to support the conversion both ways, I'm inclined to prefer not adding code we know will be dead code, wdyt?

Comment (mattiecnvr) [reply rust/gelato/src/chains.rs:61]: "Would it make sense" => No because we have no use. Fair, but was wondering if clients might want/need it since it is a lib. If not, nbd.

File: rust/gelato/src/chains.rs
Comment (mattiecnvr) [reply rust/gelato/src/chains.rs:41]: I did approve the PR. Leaving it is fine.

**Pull request merged**
